source/code/plugins/mysql_workload_lib.rb

```diff
@@ -0,0 +1,442 @@
+      @mysql ||= get_connection
+      begin
+        return @mysql.query(input_query, :cast => false, :cache_rows => false)
+      rescue Exception => e
+        @log.warn_once("mysql_workload: #{e}")
+      end
+    else
+      return @mock_interface.query(input_query)
+    end
+
+  end
+
+  def get_mysql_hostname
+    query("SHOW VARIABLES LIKE 'hostname'").each do |row|
+      return row.fetch('Value')
+    end
+  end
+
+  def transform_row(row)
+    # creates a new hash table where keys from the input row are replaced with the Variable_name key name, and the value of the key Variable_name
+    # Example {VariableName => connectedthreads, Value => 20} will be replaced with {connectedthreads = > 20}
+    result = Hash.new()
+    result.store(row["Variable_name"], row["Value"] )
+    return result
+  end
+
+  def get_value (input_array, key_to_find)
+    # finds given key in the array of hash tables and returns it's value
+    value = -1
+    # retrieves the entry value from the array of hash tables based on the given key
+    input_array.each do |row|
+      if (row.has_key?(key_to_find))
+        value = row.fetch(key_to_find)
+        break
+      end
+    end
+    return value
+  end
+
+  def get_server_database_type
+    #retrieve engine info from default_storage_engine
+    engine_query = "show variables like 'default_storage_engine'" #some versions of mysql might not populated this field so use default
+    engine_info = query(engine_query)
+    engine_info_value = "InnoDB" # default
+    engine_info.each do |row|
+      engine_info_value = row["Value"]
+    end
+    return engine_info_value
+  end
+
+  def get_server_database_version
+    #retrieve database version info frm innodb_version
+    dbversion_query = "show variables like 'version'"
+    dbversion_info = query(dbversion_query)
+    dbversion_info_value = 0 # default
+    dbversion_info.each do |row|
+      dbversion_info_value = row["Value"]
+    end
+    return dbversion_info_value
+  end
+
+  def get_server_stats
+    global_stats = Array.new
+    env_var_stats = Array.new
+    #final list of counters that will be sent to oms. Filtered based on mysql spec.
+    final_records = Array.new
+
+    # populate all mysql performance and analytical data throw SHOW GLOBAL STATUS or SHOW VARIABLES
+    q_global_status = query(@query_global_status)
+    q_global_status.each do |row|
+      new_row = transform_row(row)
+      global_stats.push(new_row)
+    end
+    q_variables = query(@query_variables)
+    q_variables.each do |row|
+      new_row = transform_row(row)
+      env_var_stats.push(new_row)
+    end
+
+    # Add NumConnections
+    num_connections = Hash.new
+    num_connections_value = get_value(global_stats, "Threads_connected")
+    num_connections.store("NumConnections", num_connections_value)
+    final_records.push(num_connections)
+
+    # Add MaxConnections
+    max_connections = Hash.new
+    max_connections_value = get_value(env_var_stats, "max_connections")
```

Comment:
Yes, there are case sensitive. the names from environment queries are lower case when global status returns upper case.